alvr/sockets/src/stream_socket/mod.rs

```diff
@@ -60,128 +61,216 @@ impl Drop for SendBufferLock<'_> {
 }
 
 impl<T> StreamSender<T> {
-    // The buffer is moved into the method. There is no way of reusing the same buffer twice without
-    // extra copies/allocations
-    pub async fn send_buffer(&mut self, mut buffer: SenderBuffer<T>) -> StrResult {
-        buffer.inner[2..6].copy_from_slice(&self.next_packet_index.to_be_bytes());
-        self.next_packet_index += 1;
-
+    pub async fn feed_socket(&mut self, packet: Bytes) -> StrResult {
         match &self.socket {
             StreamSendSocket::Udp(socket) => socket
                 .inner
                 .lock()
                 .await
-                .send((buffer.inner.freeze(), socket.peer_addr))
-                .await
-                .map_err(err!()),
-            StreamSendSocket::Tcp(socket) => socket
-                .lock()
-                .await
-                .send(buffer.inner.freeze())
+                .feed((packet, socket.peer_addr))
                 .await
                 .map_err(err!()),
+            StreamSendSocket::Tcp(socket) => socket.lock().await.feed(packet).await.map_err(err!()),
         }
     }
-}
 
-impl<T: Serialize> StreamSender<T> {
-    pub fn new_buffer(
-        &self,
-        header: &T,
-        preferred_max_buffer_size: usize,
-    ) -> StrResult<SenderBuffer<T>> {
-        let header_size = bincode::serialized_size(header).map_err(err!())?;
-        // the first two bytes are for the stream ID
-        let offset = 2 + 4 + header_size as usize;
+    pub async fn send_buffer(&mut self, buffer: &SenderBuffer<T>) -> StrResult {
+        // packet layout:
+        // [ 2B (stream ID) | 4B (packet index) | 4B (packet shard count) | 4B (shard index)]
+        // this escluses length delimited coding, which is handled by the TCP backend
+        const OFFSET: usize = 2 + 4 + 4 + 4;
+        let max_payload_size = self.max_packet_size - OFFSET;
 
-        let mut buffer = BytesMut::with_capacity(offset + preferred_max_buffer_size);
+        let buffer = buffer.inner.as_ref().unwrap();
 
-        buffer.put_u16(self.stream_id);
+        let shards = buffer.chunks(max_payload_size);
+        let shards_count = shards.len();
 
-        // make space for the packet index
-        buffer.put_u32(0);
+        let mut shards_buffer = BytesMut::with_capacity(buffer.len() + shards_count * OFFSET);
 
-        let mut buffer_writer = buffer.writer();
-        bincode::serialize_into(&mut buffer_writer, header).map_err(err!())?;
-        let buffer = buffer_writer.into_inner();
+        for (shard_index, shard) in shards.enumerate() {
+            shards_buffer.put_u16(self.stream_id);
+            shards_buffer.put_u32(self.next_packet_index);
+            shards_buffer.put_u32(shards_count as _);
+            shards_buffer.put_u32(shard_index as _);
+            shards_buffer.put_slice(shard);
 
-        Ok(SenderBuffer {
-            inner: buffer,
-            offset,
+            self.feed_socket(shards_buffer.split().freeze()).await?;
```

Comment:
As we discussed, my suggestion is to somehow match the socket here only once, as there can be many iterations matching the same socket.

Reply:
Duh, that was my intention but i lost track of the objective 🤦‍♂️

Reply:
I tried everything, boxing the futures and/or boxing closures, I always had issues with lifetimes. For now I just moved the code back to `send_buffer()`.